### ISSUE TITLE
Documentation: Ambiguity around on_text and TextEntry events

### ISSUE THREAD

**Issue by pushfoo** [pyglet/pyglet#936]
### TL;DR: The doc leaves multiple questions unanswered

1. What's the practical difference between `on_text` and `TextEntry.on_commit`?
2. Are both acceptable, or is one intended to deprecated?
3. What's the future of widgets?

### Why answer them?

We need the answers to improve `TextEntry` widgets (#931, #932, #927) and related documentation.

### What we have now

`on_text` seems to be more recent, but its example seems to start duplicating the existing widget system:
| Approach used | Example / Doc | Last Commit | 
| --- | --- | ----- |
| Override `Window.on_text` in a subclass | [`examples/text/text_input.py`](https://github.com/pyglet/pyglet/blob/master/examples/text/text_input.py) | Oct 15, 2022 | 
| Setting a handler for a `TextEntry`'s `on_commit` | [`examples/gui/widgets.py`](https://github.com/pyglet/pyglet/blob/master/examples/gui/widgets.py) | Nov 9, 2020 |

### Questions Raised

The differences between them raise the following questions:

1. Were widgets intended to be deprecated in favor of allowing users to build their own systems?
2. Is text motion / keyboard shortcut handling intended to be moved to the top level of the window?
3. Should the keyboard shortcut detection system be configurable on a per focus context / widget basis?

**Comment by pushfoo** [comment]
Reading the example history closely reveals [`text_input.py` originated in 2008](https://github.com/pyglet/pyglet/commits/03e9f4c2a8cb96576985a2e15739583090a0591c/examples/text_input.py?browsing_rename_history=true&new_path=examples/text/text_input.py&original_branch=master).

Despite its age, it appears to be more flexible than the more recent `Caret` / `TextEntry` setup due to their intertwined nature and lack of customization options:
1. There is no way to pass a `Caret` subclass to `TextEntry`
2. `Caret` does not use `EventDispatcher` as a base class, so it is not possible to override events in convenient way
3. `Caret` is marked as protected: (`TextEntry._caret`)

**Comment by caffeinepills** [comment]
To answer your questions:
> What's the practical difference between on_text and TextEntry.on_commit?
`on_commit` is simply an event fired when the user attempts to 'commit' the text. For instance, hitting enter or return after typing something in a chatbox may send (commit) the message. OR say you want to only save the changes when they 'commit'.  `on_text` is an event called by the window. You can find info [here](https://pyglet.readthedocs.io/en/latest/modules/window.html#pyglet.window.Window.on_text) about `on_text` events. 

> Are both acceptable, or is one intended to deprecated?

Both are acceptable, and used for different reasons. `text_input.py` is a demonstration on making a `IncrementalTextLayout`, which while can be a text box, can be other things. It could also be a document editor, much like notepad. It has different uses, but that is just an example. 

> Were widgets intended to be deprecated in favor of allowing users to build their own systems?

No. Widgets, from my understanding, were added as a way for users to have basic GUI elements that most applications are expected to have, such as Buttons, Sliders, and Text Input. This way a new user could make a simple menu or start screen without have to worry about making their own event handlers and logic. I personally was not part of the creation, but that's my understanding.

> Is text motion / keyboard shortcut handling intended to be moved to the top level of the window?

Those are already dispatched from the window. From what I see it goes from Window -> Frame -> Widgets.

> Should the keyboard shortcut detection system be configurable on a per focus context / widget basis?

I don't really understand. If you want to change behavior, you can subclass the WidgetBase and make your own.

I have made this more clear by updating the description of class description of the `TextEntry` widget, which should reflect on the docs next time they are built. https://github.com/pyglet/pyglet/commit/a35cd7d0ffb4165dcd4f91b68787a12ce3a905e1

**Comment by pushfoo** [comment]
Thank you for explaining thoroughly. I'll try to review the doc and code more thoroughly. I'm not yet sure, but it might be worthwhile to link your comment somewhere in the documentation if the programming guide and API doc don't yet cover everything.
